metals/src/main/scala/scala/meta/internal/metals/codeactions/CreateCompanionObjectCodeAction.scala

```diff
@@ -0,0 +1,187 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Defn
+import scala.meta.Template
+import scala.meta.Tree
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.ServerCommands
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.lsp4j.CodeActionParams
+import org.eclipse.lsp4j.Location
+import org.eclipse.{lsp4j => l}
+
+/**
+ * It creates braceless or braceful companion objects for classes, traits, and enums
+ * Then navigates to the position of the created object!
+ *
+ * @param trees
+ */
+class CreateCompanionObjectCodeAction(
+    trees: Trees,
+    buffers: Buffers
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorRewrite
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val uri = params.getTextDocument().getUri()
+
+    val path = uri.toAbsolutePath
+    val range = params.getRange()
+    val applyTree =
+      if (range.getStart == range.getEnd)
+        trees
+          .findLastEnclosingAt[Tree](
+            path,
+            range.getStart(),
+            applyWithSingleFunction
+          )
+      else
+        None
+
+    applyTree.flatMap { tree =>
```

Comment:
If we used Option for `getNameOfClassTraitOrEnumTree` we can rewrite this all to:
```scala
val companionObjectAction = for {
  tree <- applyTree
  name <- getNameOfClassTraitOrEnumTree(tree)
  if !hasCompanionObject(tree, name)
  document <- buffers.get(path).getOrElse("")
} yield buildCreatingCompanionObjectCodeAction(
            path,
            tree,
            uri,
            indentationString,
            name,
            hasBraces(tree, document)
          )
companionObjectAction.toSeq
```

We often use for comprehensions in the codebase as it makes it a bit clearer when we have Options